code-studio/src/js/pd/workshop_dashboard/components/workshop_form.jsx

```diff
@@ -0,0 +1,567 @@
+    return (
+      <Input
+        type="select"
+        label="Workshop Type"
+        value={this.state.workshop_type || ''}
+        onChange={(event) => {this.handleFieldChange('workshop_type', event.target.value);}}
+        bsStyle={validation.style.workshop_type}
+        help={validation.help.workshop_type}
+        style={this.props.readOnly && styles.readOnlyInput}
+        disabled={this.props.readOnly}
+      >
+        {placeHolder}
+        {options}
+      </Input>
+    );
+  },
+
+  shouldRenderSubject() {
+    return this.state.course && WORKSHOP_CONSTANTS.SUBJECTS[this.state.course];
+  },
+
+  renderSubjectSelect: function (validation) {
+    if (this.shouldRenderSubject()) {
+      var options = WORKSHOP_CONSTANTS.SUBJECTS[this.state.course].map(function (subject, i) {
+        return (<option key={i} value={subject}>{subject}</option>);
+      });
+      var placeHolder = this.state.subject ? null : <option />;
+      return (
+        <Input
+          type="select"
+          label="Subject"
+          value={this.state.subject || ''}
+          onChange={(event) => {this.handleFieldChange('subject', event.target.value);}}
+          bsStyle={validation.style.subject}
+          help={validation.help.subject}
+          style={this.props.readOnly && styles.readOnlyInput}
+          disabled={this.props.readOnly}
+        >
+          {placeHolder}
+          {options}
+        </Input>
+      );
+    } else {
+      this.state.subject = null;
+    }
+  },
+
+  handleErrorClick: function (i) {
+    this.state.errors.splice(i,1);
+    this.setState(this.state);
+  },
+
+  renderErrors: function () {
+    if (!this.state.errors || this.state.errors.length == 0) {
+      return null;
+    }
+    return this.state.errors.map(function (error, i) {
+      return (
+        <Alert
+          bsStyle="danger"
+          key={i}
+          onDismiss={this.handleErrorClick.bind(null, i)}
+        >
+          {error}
+        </Alert>
+      );
+    }.bind(this));
+  },
+
+  shouldConfirmSave: function () {
+    if (!this.props.workshop) {
+      return false;
+    }
+    return (
+      this.state.sessionsModified ||
+      this.state.location_name != this.props.workshop.location_name ||
+      this.state.location_address != this.props.workshop.location_address ||
+      this.state.notes != this.props.workshop.notes
+    );
+  },
+
+  handleSaveClick: function (e) {
+    var validation = this.validate();
+    if (validation.isValid) {
+      if (this.shouldConfirmSave()) {
+        this.state.showSaveConfirmation = true;
+        this.setState(this.state);
+      } else {
+        this.save(false);
+      }
+    } else {
+      this.state.shouldValidate = true;
+      this.setState(this.state);
+    }
+  },
+
+  handleSaveAndNotifyClick: function () {
+    this.save(true);
+  },
+
+  handleSaveNoNotifyClick: function () {
+    this.save(false);
+  },
+
+  handleAbortSave: function () {
+    this.state.showSaveConfirmation = false;
```

Comment:
Directly modifying this.state. I'm going to stop calling each of these out, because it will probably be more noisy than helpful to you. But I do think it's important that we fix each of these